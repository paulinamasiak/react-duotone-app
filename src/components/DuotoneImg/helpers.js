const HEX_COLOR_REGEXP = /^#([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i;

// supports only normal hex color notation
export const hexToRgb = (hex) => {
  const result = HEX_COLOR_REGEXP.exec(hex);

  if (!result) {
    return null;
  }

  return {
    r: parseInt(result[1], 16),
    g: parseInt(result[2], 16),
    b: parseInt(result[3], 16),
  };
}

export const gradientMap = (tone1, tone2) => {
  const rgb1 = hexToRgb(tone1);
  const rgb2 = hexToRgb(tone2);
  const gradient = [];

  for (let i = 0; i < (256 * 4); i += 4) {
    gradient[i] = ((256 - (i / 4)) * rgb1.r + (i / 4) * rgb2.r) / 256;
    gradient[i + 1] = ((256 - (i / 4)) * rgb1.g + (i / 4) * rgb2.g) / 256;
    gradient[i + 2] = ((256 - (i / 4)) * rgb1.b + (i / 4) * rgb2.b) / 256;
    gradient[i + 3] = 255;
  }

  return gradient;
};
