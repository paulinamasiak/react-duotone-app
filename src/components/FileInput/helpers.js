import { isEmpty } from 'lodash';

export const isAcceptedFile = (file, accept) => {
  if (isEmpty(accept)) {
    return true;
  }

  if (isEmpty(file)) {
    return false;
  }

  const acceptedTypes = accept.split(',');
  const fileName = file.name;
  const fileType = file.type;

  return acceptedTypes.some((type) => {
    const validType = type.trim();

    if (validType.startsWith('.')) {
      return fileName.endsWith(validType);
    }

    if (validType.endsWith('/*')) {
      return fileType.startsWith(validType.replace(/\/.*$/, ''));
    }

    return fileType === validType;
  });
};
