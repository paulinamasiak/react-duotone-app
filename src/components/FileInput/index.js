import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { noop } from 'lodash';
import { isAcceptedFile } from './helpers';

class FileInput extends Component {
  constructor(props) {
    super(props);

    this.file = null;
    this.state = {
      isDragOver: false,
    };

    this.handleDragEnter = this.handleDragEnter.bind(this);
    this.handleDragOver = this.handleDragOver.bind(this);
    this.handleDragLeave = this.handleDragLeave.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleDragEnter() {
    this.setState({
      isDragOver: true,
    });
  }

  handleDragOver(event) {
    event.preventDefault();
  }

  handleDragLeave() {
    this.setState({
      isDragOver: false,
    });
  }

  handleDrop(event) {
    event.preventDefault();

    const file = event.dataTransfer.files[0];
    const { accept } = this.props;

    if (isAcceptedFile(file, accept)) {
      this.file = file;
      this.props.onChange(this.file);
    }

    this.setState({
      isDragOver: false,
    });
  }

  handleChange(event) {
    this.file = event.target.files[0];
    this.props.onChange(this.file);
  }

  render() {
    const className = classNames('file-input', {
      'is-dragover': this.state.isDragOver,
    });

    return (
      <label
        className={className}
        onDragEnter={this.handleDragEnter}
        onDragOver={this.handleDragOver}
        onDragLeave={this.handleDragLeave}
        onDrop={this.handleDrop}
      >
        <input
          type="file"
          accept={this.props.accept}
          onChange={this.handleChange}
        />
        <span className="file-input-dnd">{this.props.placeholder}</span>
        <i className="fas fa-plus" />
      </label>
    );
  }
}

FileInput.propTypes = {
  accept: PropTypes.string,
  placeholder: PropTypes.string,
  onChange: PropTypes.func,
};

FileInput.defaultProps = {
  accept: '',
  name: '',
  placeholder: '',
  onChange: noop,
};

export default FileInput;
