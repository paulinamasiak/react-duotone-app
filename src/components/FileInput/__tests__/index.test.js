import React from 'react';

import { shallow } from 'enzyme';
import { isAcceptedFile } from '../helpers';
import FileInput from '../index';

jest.mock('../helpers');

describe('FileInput component', () => {
  it('should render without crashing', () => {
    // when
    const wrapper = shallow(<FileInput />);

    // then
    expect(wrapper.exists()).toBe(true);
  });

  it('should render correctly with default props', () => {
    // when
    const wrapper = shallow(<FileInput />);

    // then
    expect(wrapper.getElements()).toMatchSnapshot();
  });

  it('should display placeholder based on props.placeholder', () => {
    // when
    const wrapper = shallow(<FileInput placeholder="Drag & drop a file" />);

    // then
    expect(wrapper.find('.file-input-dnd').text()).toBe('Drag & drop a file');
  });

  it('should pass props.accept to the input accept prop', () => {
    // given & when
    const wrapper = shallow(<FileInput accept="image/*" />);

    // then
    expect(wrapper.find('input').prop('accept')).toBe('image/*');
  });

  describe('on dragEnter', () => {
    it('should add is-dragover class', () => {
      // given
      const wrapper = shallow(<FileInput />);

      // when
      wrapper.simulate('dragEnter');

      // then
      expect(wrapper.hasClass('is-dragover')).toBe(true);
    });
  });

  describe('on dragOver', () => {
    it('should call preventDefault', () => {
      // given
      const wrapper = shallow(<FileInput />);
      const mockEvent = {
        preventDefault: jest.fn(),
      };

      // when
      wrapper.simulate('dragover', mockEvent);

      // then
      expect(mockEvent.preventDefault).toHaveBeenCalled();
    });
  });

  describe('on dragLeave', () => {
    it('should remove is-dragover class', () => {
      // given
      const wrapper = shallow(<FileInput />);

      // when
      wrapper.simulate('dragEnter');
      wrapper.simulate('dragLeave');

      // then
      expect(wrapper.hasClass('id-dragover')).toBe(false);
    });
  });

  describe('on drop', () => {
    let wrapper;
    let onChangeSpy;
    let eventMock;

    beforeEach(() => {
      // given
      onChangeSpy = jest.fn();
      wrapper = shallow(<FileInput onChange={onChangeSpy} />);
      eventMock = {
        preventDefault: jest.fn(),
        dataTransfer: {
          files: [{ name: 'image.jpg', type: 'image/jpeg' }],
        },
      };

      wrapper.simulate('dragEnter');
    });

    it('should call preventDefault', () => {
      // when
      wrapper.simulate('drop', eventMock);

      // then
      expect(eventMock.preventDefault).toHaveBeenCalled();
    });

    it('should call props.onChange when dropped file is accepted', () => {
      // given
      isAcceptedFile.mockReturnValue(true);

      // when
      wrapper.simulate('drop', eventMock);

      // then
      expect(onChangeSpy).toHaveBeenCalledWith({
        name: 'image.jpg',
        type: 'image/jpeg',
      });
    });

    it('should not call props.onChange when file is not accepted', () => {
      // given
      isAcceptedFile.mockReturnValue(false);

      // when
      wrapper.simulate('drop', eventMock);

      // then
      expect(onChangeSpy).not.toHaveBeenCalled();
    });

    it('should remove is-dragover class', () => {
      // when
      wrapper.simulate('drop', eventMock);

      // then
      expect(wrapper.hasClass('is-dragover')).toBe(false);
    });
  });

  describe('on input change', () => {
    it('should call props.onChange with a single file', () => {
      // given
      const onChangeSpy = jest.fn();
      const wrapper = shallow(<FileInput onChange={onChangeSpy} />);
      const mockEvent = {
        target: {
          files: [{ name: 'image.jpg', type: 'image/jpeg' }],
        },
      };

      // when
      wrapper.find('input').simulate('change', mockEvent);

      // then
      expect(onChangeSpy).toHaveBeenCalledWith({
        name: 'image.jpg',
        type: 'image/jpeg',
      });
    });
  });
});
