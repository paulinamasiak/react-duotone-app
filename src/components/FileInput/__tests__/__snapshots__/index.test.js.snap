// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput component should render correctly with default props 1`] = `
Array [
  <label
    className="file-input"
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
>
    <input
        accept=""
        onChange={[Function]}
        type="file"
    />
    <span
        className="file-input-dnd"
    />
    <i
        className="fas fa-plus"
    />
</label>,
]
`;
